timemachine/cpp/src/kernels/k_barostat.cuh

```diff
@@ -0,0 +1,153 @@
+    const int *__restrict__ atom_idxs,         // [N]
+    const int *__restrict__ mol_idxs,          // [N]
+    unsigned long long *__restrict__ centroids // [num_molecules * 3]
+) {
+    int idx = blockIdx.x * blockDim.x + threadIdx.x;
+    while (idx < N) {
+        int atom_idx = atom_idxs[idx];
+        int mol_idx = mol_idxs[idx];
+        atomicAdd(centroids + mol_idx * 3 + 0, FLOAT_TO_FIXED<double>(coords[atom_idx * 3 + 0]));
+        atomicAdd(centroids + mol_idx * 3 + 1, FLOAT_TO_FIXED<double>(coords[atom_idx * 3 + 1]));
+        atomicAdd(centroids + mol_idx * 3 + 2, FLOAT_TO_FIXED<double>(coords[atom_idx * 3 + 2]));
+        idx += gridDim.x * blockDim.x;
+    }
+}
+
+void __global__ k_setup_barostat_move(
+    const double *__restrict__ rand,     // [2], use first value, second value is metropolis condition
+    double *__restrict__ d_box,          // [3*3]
+    double *__restrict__ d_volume_delta, // [1]
+    double *__restrict__ d_volume_scale, // [1]
+    double *__restrict__ d_length_scale  // [1]
+) {
+    const int idx = blockIdx.x * blockDim.x + threadIdx.x;
+    if (idx >= 1) {
+        return; // Only a single thread needs to perform this operation
+    }
+    const float volume = d_box[0 * 3 + 0] * d_box[1 * 3 + 1] * d_box[2 * 3 + 2];
+    if (d_volume_scale[0] == 0) {
+        d_volume_scale[0] = 0.01 * volume;
+    }
+    const float delta_volume = d_volume_scale[0] * 2 * (rand[0] - 0.5);
+    const float new_volume = volume + delta_volume;
+    d_volume_delta[0] = delta_volume;
+    d_length_scale[0] = cbrt(new_volume / volume);
+}
+
+void __global__ k_decide_move(
+    const int N,
+    const int num_molecules,
+    const double kt,
+    const double pressure,
+    const double *__restrict__ rand, // [2] Use second value
+    double *__restrict__ d_volume_delta,
+    double *__restrict__ d_volume_scale,
+    const __int128 *__restrict__ d_init_u,
+    const __int128 *__restrict__ d_final_u,
+    double *__restrict__ d_box,
+    const double *__restrict__ d_box_output,
+    double *__restrict__ d_x,
+    const double *__restrict__ d_x_output,
+    int *__restrict__ num_accepted,
+    int *__restrict__ num_attempted) {
+    int idx = blockIdx.x * blockDim.x + threadIdx.x;
+
+    const double volume = d_box[0 * 3 + 0] * d_box[1 * 3 + 1] * d_box[2 * 3 + 2];
+    const double volume_delta = d_volume_delta[0];
+    const double new_volume = volume + volume_delta;
+    double energy_delta = INFINITY;
+    if (!fixed_point_overflow(d_final_u[0]) && !fixed_point_overflow(d_init_u[0])) {
+        energy_delta = FIXED_ENERGY_TO_FLOAT<double>(d_final_u[0] - d_init_u[0]);
+    }
+
+    const double w = energy_delta + pressure * volume_delta - num_molecules * kt * std::log(new_volume / volume);
+
+    const bool rejected = w > 0 && rand[1] > std::exp(-w / kt);
+
+    while (idx < N) {
+        if (idx == 0) {
+            if (!rejected) {
+                num_accepted[0]++;
```

Comment:
nit/curiosity: is there a reason to prefer the `num_accepted[0]` dereference syntax over `*num_accepted`? (was initially confused thinking this was an array)

Reply:
I interpreted this style to be the convention in the repo.

Reply:
I don't think we standardized on this - no opinion tbh